### DISCUSSION TITLE
AI Green Waste Project

### DISCUSSION THREAD

**Discussion by etoschi** [c0c0n3/kitt4sme.live#141] [Q&A] [answered]
Q: the picture attached summarizes the assumptions for what concerns the development and deployment setup for the **AI Green Waste Project**. Namely:

application source code is kept on a versioning server in corporate servers; no need to push this on a github repository
when developing application code, the developers push their changes to the corporate git, then build it into one or more docker images; the resulting images are pushed to a docker repository – steps (1), (2) and (3) in the figure below.
the code for kubernetes descriptors and configuration for deployment on kitt4sme platform is kept in a separate git repository; the origin for this repository is a private fork on github of the kitt4sme.live github project; at this stage we’re not pushing our code to the main copy of the github repo (c0c0n3/kitt4sme.live), but on our own private fork, consequently components are not deployed on the online platform, but will be deployed on our own instance of the platform installed on corporate servers (bottom-right box in the picture). – these are depicted as steps (4), (5) and (6) on the picture.
At the end of the project (when things are stable), we’ll be able to push/deploy our code on kitt4sme’s online platform.
On our own instance of the platform, we will configure ArgoCD to pull changes from our private fork of kitt4sme.live and images from the docker repository, and to deploy components on the same platform instance.
Cuold you please confirm that the above assumptions are correct?

 

The following questions are based on the assumptions above – in case they are already answered somewhere in the documentation, could you please just point to the correct page url?

 

Q: Which docker registry should we use for deployment? Shall we push our images on docker hub, or is there some docker registry specifically setup for kitt4sme? In such case, where do we find connection details? How do we configure ArgoCD to grab images from such a registry? (didn’t find in the documentation)

 

Q: When installing the demo cluster on our servers, the [bootstrap guide](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/bootstrap.md), in post install steps, announces that “the only unhappy service should be keycloak”, as it actually happens, with the keycloak pod stuck in a CreateContainerConfigError. The guide then points to the  [security how-to](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/security.md) document. But there I couldn’t find any instruction to get keycloak running. Actually, going to http://<SERVER_IP>/auth in the browser I get a “no healthy upstream” message. In the pod’s describe output, among the events there is:

Warning – Failed - 55m (x4 over 55m) - kubelet  Error: secret "keycloak-builtin-admin" not found

The keycloak-builtin-admin.yaml file is present in deployment\mesh-infra\security\secrets

 

Q: the Orion broker seems to be running on the platform; does it have some UI or dashboard link to access it to be able to check that data is correctly flowing to it?
[kitt4sme demo cluster technical questions.docx](https://github.com/c0c0n3/kitt4sme.live/files/8563601/kitt4sme.demo.cluster.technical.questions.docx)

**Comment by c0c0n3** [comment]
Hi @etoschi! Thanks for asking these questions I think the discussion will be relevant for other open call winners too. I'm attaching here the diagram you put in the Word doc so it's easier to see. Nice diagram by the way, thank you so much for sharing!

![algowatt](https://user-images.githubusercontent.com/13233181/165460933-e7eb4dbd-2328-4d6f-b343-37c08bf48fbc.png)

**Comment by c0c0n3** [comment, marked as answer]
On to your questions now :-)

### Fork workflow & contribs

> application source code is kept on a versioning server in corporate servers...
> ... from the docker repository, and to deploy components on the same platform instance.
> Cuold you please confirm that the above assumptions are correct?

Yes, totally what we had in mind. Thank you so much for writing this up, it'll be useful as a reference for other open call winners too.

### Docker registry

> Which docker registry should we use for deployment?
> Shall we push our images on docker hub, or is there some docker registry specifically setup for kitt4sme? 

Use DockerHub for the time being---or any other public registry. Going forward we'll migrate images to the RAMP registry but at the moment there are still some details we need to work out.

> In such case, where do we find connection details?

We'll publish a little guide on how to work with the RAMP registry when the time to switch over comes.

> How do we configure ArgoCD to grab images from such a registry? (didn’t find in the documentation)

ArgoCD isn't pulling images. K8s drives that. If the image sits in a pub registry then you just need to reference your image in the K8s manifest image field. If the image is in a private registry, then you've got to create a secret too for K8s to be able to pull it:

- https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/

Secrets should be managed as explained in our [how-to][sec.how-to].

### Keycloak bootstrap

> When installing the demo cluster on our servers, the [bootstrap guide](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/bootstrap.md), in post install steps, announces that “the only unhappy service should be keycloak”, as it actually happens, with the keycloak pod stuck in a CreateContainerConfigError. The guide then points to the [security how-to](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/security.md) document. But there I couldn’t find any instruction to get keycloak running. Actually, going to http://<SERVER_IP>/auth in the browser I get a “no healthy upstream” message. In the pod’s describe output, among the events there is:
>
> Warning – Failed - 55m (x4 over 55m) - kubelet Error: secret "keycloak-builtin-admin" not found
>
> The keycloak-builtin-admin.yaml file is present in deployment\mesh-infra\security\secrets

I think the problem is that you've got to regenerate the `keycloak-builtin-admin.yaml` file and then push it to the repo. Try following the steps in the "Managing secrets / Workflow" section of the [security how-to][sec.how-to]. The docs aren't great and assume you're familiar with K8s secrets, Sealed Secrets and Reloader. If you can't get it working, get in touch we'll try scheduling a session to help you out with the setup :-)

### Orion UI

> the Orion broker seems to be running on the platform;

Awesome, well done!

> does it have some UI or dashboard link to access it to be able to check that data is correctly flowing to it?

Unfortunately Orion doesn't ship with a UI or dashboard. But if you just want to check visually your requests are hitting the service, you could use the Kiali dashboard:

```
$ cd nix
$ nix shell
$ export KUBECONFIG=/path/to/your/kube/config
$ istioctl d kiali
```

[You can display a service graph][kiali] showing requests flowing to any service in real time. For a bit more detail you can navigate to the Orion app and have a look at the log lines Kiali reads in real time from the service. There you should be able to see Orion logging the processing of your requests as they get through. (Logs are also available through the ArgoCD UI.) Also, you could use Postman to query Orion---e.g. get entities you posted earlier.

There are also plenty of monitoring options through the command line---e.g. `kubectl`, `mongo shell`, etc. But I reckon I'm drifting off-topic since you asked for visual tools :-)



[sec.how-to]: https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/security.md
[kiali]: https://istio.io/latest/docs/tasks/observability/kiali/

**Reply by etoschi** [reply]
Thanks!

**Comment by c0c0n3** [comment]
pleasure!
